### ISSUE TITLE
Skipping frames / issue with savedinstances update

### ISSUE THREAD

**Issue by sean2648** [eTzmNcbkrng/SavedInstances#38]
Downloaded the github update for savedinstances and now my game basically has a hiccup or skips every 5 seconds or so. Turn off all addons except savedinstances to reproduce the problem. If I use the old version from twitch app or curse the problem doesn't occur. Only happens with this new version.

Any help or suggestions would be appreciated as I like the mythic keystone functionality and that doesn't exist with the old version. But the game is almost unplayable with this new version installed for whatever reason.

Thanks

**Comment by sean2648** [comment]
yyo6
View User Profile
Send Message
Follow User
Ignore User
Posted 57 mins ago
I got exactly the same issue and aswell old version works just fine without stuttering. Please anybody can help?


Another user under the comments on curse with the same issue. Just wanted to post as follow up. Really love the addon and would like to be able to use it again.

**Comment by Rubeek** [comment]
Exact same issue here too. Very annoying, ended up disabling the add-on for now. Hopefully this gets fixed.

**Comment by Stanzilla** [comment]
Won't be near a WoW capable PC for the next few weeks. Did you guys try older versions yet? Was it a SI change that broke it or the WoW patch?

**Comment by Rubeek** [comment]
So this is odd.. While trying out older versions to see where the issue started to occur I deleted my SavedInstances folder from my AddOns folder to completely replace it and now it appears I'm having no issues with any version, including the update that went out around 17 hours ago. I checked each version from Sept 6th and they were all without issue. I'm not sure if deleting the original folder had anything to do with fixing my issue but I imagine it played some part. I probably should have tried this first before complaining but seeing other people with the same(or similar) issue I assumed it wasn't my end, sorry!

If you haven't already, I'd recommend trying this @sean2648, maybe it will help.

**Comment by Stanzilla** [comment]
Glad to see it works now for you, some of this code is very old needs to be cleaned up some more so it is certainly possible that a fresh profile fixes things for people.

**Comment by sean2648** [comment]
deleted the savedinstances folder, deleted the saved variables and backup for the actual addon and then got the most recent version. that fixed everything it seems.

I was just using an older version that was working before trying this.

**Comment by rubyrabble** [comment]
I tried what sean2648 suggested and it did fix it. But I also had the lag/choppy play come back; seemingly after all my characters had been logged in the addon had been updated with their information.

**Comment by iMorgenstjerne** [comment]
I love this addon, but I'm also having this issue 👎 I'll keep my eye on this thread to see if theres a fix for it. Fingers crossed!

**Comment by Stanzilla** [comment]
Are you guys getting any errors? Are Lua errors turned on? `/console scriptErrors 1`

**Comment by iMorgenstjerne** [comment]
I’m not getting any LUA errors, just hickups/framedrops every couple seconds. I’ve just installed the newest version that got pushed just now, so far so good. I’ll report back 😁

**Comment by sean2648** [comment]
No LUA errors for me either, and yes they are turned on.

Still getting the hiccups / dropped frames / skipping. It happens roughly every 4-5 seconds. I have a feeling most people are oblivious to it happening to them or blame it on blizzard / their computers. Or perhaps they haven't narrowed down the addon. 

Right now I am using an old version that doesn't seem to have messed up yet. It was on a github by a different person.

**Comment by Stanzilla** [comment]
I assume you guys have some sort of m+ stone on some char and that it is related to that. I have no stone on this account atm and everything is fine.

**Comment by SiegaPlays** [comment]
I have not been doing M++ for a while and i seem to have no problem,or maybe I am totally oblivious to even notable stutters - in which case I should stop on the coffee and get more sleep.

**Comment by sean2648** [comment]
got the problem again last night, had to just get rid of the addon

**Comment by sean2648** [comment]
Would really like to get some steam on getting this fixed, love the addon.

**Comment by Stanzilla** [comment]
Go test every commit then, finding where it broke is key.

**Comment by dsabecky** [comment]
I'm not having this problem. Smooth 60fps+ constantly.

**Comment by Daeveren** [comment]
After selectively disabled/enabled my addons, I found that this fork of SavedVariances was the culprit for freezes every 5 seconds or so.
I have deleted both the addon folder and the saved variables and it does the same.
I am willing to test multiple commits, to find the one that broke it!

**Comment by Daeveren** [comment]
Right, I have pinpointed it to the Sep 8, 2017 commit as being the culprit!
So basically how I tested (deleted addon folder, and saved variable before each new attempt) was:
Without SavedInstances loaded: constant 60 fps ingame using the default fps counter
With SavedInstances older commits (tried several commits between 5 aug - 6 sep), ingame fps drops to 56 every ~5 seconds.
With SavedInstances commits from 8 sep up to 1 oct one, ingame fps drops to 45 every 5 seconds and this freeze is really noticeable.
I'm going to test it further with older versions, prior to 5 aug, because I don't like that drop to 56 fps either and it definatelly shouldn't happen, there must be something wrong somewhere before that commit too.

**Comment by SacredDuckwhale** [comment]
This is the commit in question: 0af827a18effc61082db5cd4146a978a5d8d899f

SavedInstances has registered for several events that basically spam-fire even if the emissary cache can't possibly have updated. In my tests, each execution only takes ~7ms though so I am not sure if that can really cause FPS issues? My system is pretty solid, so maybe it doesn't affect me as much.

---

Actually, I noticed that the addon causes some unnecessary garbage and that could sometimes cause very minor FPS drops. With all addons disabled, it went from 100 FPS to 99 or 98 most of the time, but every once in a while it'd be 94 or so.
I don't know if this would be different on weaker systems, but it seems to have occured when SI executed that function, . However, since it doesn't *always* happen and there may be other causes, this alone doesn't mean anything really.

My best guess at this time is that the frequent accrual of garbage causes (additional) garbage collection cycles, which can indeed affect performance under certain circumstances. The FPS drops coincide with those cycles, as is evident from monitoring the memory usage: >5 FPS drops usually happened only when the addon's memory was reduced visibly, by some hundred kilobytes.

---

If anyone wants to test it, what I did was modify the respective function like so for a very basic profiling approach (can't upvalue the functions, but it shouldn't affect performance too much...): [GitHub gist](https://gist.github.com/SacredDuckwhale/81dd38a0327dc50651d6aa112bb7a7c1)

I simply disabled all addons, went to a quiet place (graphics settings lowered also), and watched the Game Menu's info (FPS and memory usage) by hovering over the "?" icon in the menu. Feel free to test this to see if you can confirm the issue :P

**Comment by Stanzilla** [comment]
I reverted that commit already since it was not actually needed, still agree on performance concerns but I won't have the time to look into it any time soon.
